src/allmydata/web/status.py

```diff
@@ -1140,6 +1155,50 @@ def render_JSON(self, req):
 
         return json.dumps(data, indent=1) + "\n"
 
+    def getChild(self, path, request):
+        # The "if (path is empty) return self" line should handle
+        # trailing slash in request path.
+        #
+        # Twisted Web's documentation says this: "If the URL ends in a
+        # slash, for example ``http://example.com/foo/bar/`` , the
+        # final URL segment will be an empty string. Resources can
+        # thus know if they were requested with or without a final
+        # slash."
+        if not path:
+            return self
+
```

Comment:
In another PR, I think @exarkun pointed out that this will make any number of trailing slashes value (e.g. `/foo/bar////`)

Reply:
Hmm, that is correct!  `$WUI/status/////////////` also is handled by this class.  I don't know if that is a cause for concern though.

Reply:
The right behavior is a 404 response.  `$WUI/status/` means *the "" child of the "status" child of the WUI*.  `$WUI/status//` means *the "" child of the "" child of the "status" child of the WUI*.  There may be a case for status having a "" child but I don't think there's any case for that resource having a "" child.

In general we shouldn't be lenient in accepting extra /s in URLs because being lenient in what you accept like this is fertile ground for security issues.

However, I've lost track of how we're doing on this front overall.  The issue seems to come up again and again in the web code.  I'm not clear on whether this is because of some mismatch between Nevow and Twisted Web or if it's a result of decisions made in Tahoe-LAFS or if maybe there is no general explanation and there's just a lot of unique cases.

This case seems to be due to `allmydata.test.web.test_web.Web.test_status`?  That is the only test I see fail if I comment out these two lines.  And it appears to fail because it wants to be able to request `/status/?t=json`.  Why does it do that?  I don't know, the test has no docstring and it tests half a dozen separate things so it's not exactly straightforward to follow.  The same test also makes a request for `/status`!  So I suppose the test essentially believes that `/status` and `/status/` should be identical.  This is poor interface design since the URIs point to resources at different depths into the hierarchy and so it's hard to generate correct links to place on the resulting page.

The frontend docs *only* talk about `/status/` and not `/status` so it's not clear why the test exercises them both.  Who's wrong, the docs or the test/implementation?

Okay.  The test was introduced in 914783acdf271a4f59967b58767827b5775da68b and it didn't change the behavior of `/status` or `/status/` at all beyond spitting out json when `t=json` is included.  This makes me think the trailing / was an accident or a misunderstanding.  But looking at this diff also reminds me that the Nevow behavior comes from `addSlash` and that behavior is to generate a redirect from `/status` to `/status/` automatically.

Reading about `addSlash` makes me think we have a systematic issue here where that Nevow feature was widely used in Tahoe-LAFS and Twisted Web doesn't even have anything like it.  Perhaps we could add this functionality in some general/reusable way so all the Tahoe-LAFS web code can benefit from it.  

Considering the old behavior was to generate a redirect from `/status` to `/status/` that seems like it's the ideal behavior to present going forward.  This can be implemented in a somewhat straightforward and reusable manner.  

We can have a `AddSlash` `IResource` wrapper.  `AddSlash.render` should generate a redirect to the request URI + "/".  `AddSlash.getChildWithDefault("", request)` should return the wrapped resource.  `AddSlash.getChildWithDefault(anything_else, request)` should delegate to the wrapped resource's `getChildWithDefault`.

Then any resource we have where we want the same content at /foo and /foo/ and for which a redirect to the /-version is okay, we can add it to the hierarchy wrapped in `AddSlash` (so, eg, `self.putChild(b"status", AddSlash(Status(...)))` in `root.py`).

Long term, we probably want to do away with all these URI aliases.  We could also consider just starting that process now and breaking the non-slash version of this URI.  We just need to document it and update the tests.

I'm inclined to support it a while longer and give people a smoother transition but if the above `AddSlash` stuff seems like too much work I'm not going to argue much for it.

Reply:
I don't have a considered opinion on this.  I don't know what the relevant standards are and what they say about this.  It seems that there's no broad consensus among implementations on how to handle trailing slashes.

For example, a `GET` on `https://tahoe-lafs.org/trac/tahoe-lafs` and `https://tahoe-lafs.org/trac/tahoe-lafs/` and `https://tahoe-lafs.org/trac/tahoe-lafs//` and `https://tahoe-lafs.org/trac/tahoe-lafs////`  all get almost identical responses.

Same with several popular websites I tested too, except that the tracking junk changes with each request.

Haskell [Warp](https://hackage.haskell.org/package/warp) (or maybe it is WAI or Yesod) treats `GET http://https://snowdrift.coop` and `GET https://snowdrift.coop/` and `GET https://snowdrift.coop//` the same, but `GET https://snowdrift.coop///` gets a `301 Moved Permanently`.  But it's a little different when there's a subpath: `GET https://snowdrift.coop/projects` gets a `200` response; `GET https://snowdrift.coop/projects/` gets a `301 Moved Permanently` to `Location: https://snowdrift.coop/projects`.

Seems that Django websites are set up to return `404 Not Found` when there are extra trailing slashes.  And it seems that Ruby on Rails and Phoenix do not care for the trailing slashes.

I am not saying that any of them are doing the right thing... just that this is an interestingly quirky thing. :-)